### ISSUE TITLE
PostgreSQL 9.5 IMPORT FOREIGN SCHEMA

### ISSUE THREAD

**Issue by robe2** [pramsey/pgsql-ogr-fdw#10]
This is not a bug, but a feature request.  

http://michael.otacoo.com/postgresql-2/postgres-9-5-feature-highlight-import-foreign-schema/

**Comment by robe2** [comment]
I've started playing around with this.  It's not that much work I don't think.  A lot of the logic @pramsey  has in ogrfdw_info can be repurposed for this and I've already cut in the IMPORT FOREIGN schema glue I saw in the postgresql_fdw code.

One thing I'm not sure how to handle is IMPORT FOREIGN SCHEMA <remote_schema> 

the  <remote_schema>  is a required bit.  There are two ways that come to mind to handle this:

1) come up with some bogus remote schema like 'all'  that would mean ignore the schema and that people would have to specify

2) just ignore the schema if the data source does not support schemas

 For schema checking I was going to assume that databases that have schema's the tables come thru via with dot's


For example when I connect to SQL Server, table names come down as

```

dbo.Orders
dbo.Products

```

etc. but not sure if that is a true statement for all/most OGR data sources or if there is a logic in OGR to handle schemas.  I don't think there is.

The other minor issue is handling EXCEPT and LIMIT TO clauses in IMPORT FOREIGN SCHEMA.

One benefit about using a specific schema word to mean ignore schemas, is that if a true schema name is used, then we no the EXCEPT / LIMIT TO need not include the schema name (or must in the case of all).

**Comment by robe2** [comment]
Okay getting warm now, I have successfully imported a table with import foreign schema.  Now to work out some kinks.

It does seem that there is a mechanism for options:

http://www.postgresql.org/docs/9.5/static/sql-importforeignschema.html

So for first option, I wanted to define a launder column names option.  If that is not set to true, then the column names would come in exactly as they are in remote data source, instead of being forced to lower case and having unsavory names changed. Not sure if the launder should apply to table name as well (minus the schema part) or if we should have a separate option for that.

**Comment by robe2** [comment]
Okay I decided to go with two separate laundering options and default both to true if not specified.

So I have:

```  
launder_table_names
launder_column_names

```

So to use for example I have a statements like this:

```
CREATE SERVER svr_legacy_btu
   FOREIGN DATA WRAPPER ogr_fdw
  OPTIONS (datasource 'C:\data\BTUData.mdb',format 'ODBC');

DROP SCHEMA IF EXISTS btu CASCADE;
CREATE SCHEMA btu;

IMPORT FOREIGN SCHEMA public 
FROM server svr_legacy_btu INTO btu OPTIONS(launder_table_names 'false') ;

```
Since I set it to false, the casing and weird junk in my table names are preserved.

Similarly if I do this:

````

DROP SCHEMA IF EXISTS btu CASCADE;
CREATE SCHEMA btu;

IMPORT FOREIGN SCHEMA public 
FROM server svr_legacy_btu 
INTO btu OPTIONS(launder_table_names 'false', launder_column_names 'false') ;

```
Both table and column name casing and weird characters are preserved.  Now all I have left is to tackle the EXCEPT / LIMIT TO and do something with the remote schema and I think my patch (when I submit it - after @pramsey has committed my 9.5rc1 fix patch) will be feature complete :).

P.S. I also noticed for some reason I have to use a schema name that exists in my db for the remote schema, which seems pretty lame.  Not sure this is a bug or just a limitation of existing IMPORT..functionality.

**Comment by robe2** [comment]
For remote schema functionality, I decided to go with using the remote_schema as a prefix instead of as a real schema.  Reason is things like WFS could have very nested names with periods and the concept then also becomes useful for things that don't have schemas.

For the ignore schema option, I decided to go with name <b>ogr_all</b> for remote schema name.

So for example:

```
CREATE SERVER fgdbtest
  FOREIGN DATA WRAPPER ogr_fdw
  OPTIONS (
    datasource 'c:/projects/sources/pgsql_ogr_fdw_robe/data/Querying.gdb',
    format 'OpenFileGDB' );

DROP SCHEMA IF EXISTS fgdbq CASCADE;
CREATE SCHEMA fgdbq;

-- will just import CitiesInfo  table (basically all tables starting with CitiesInf) --
IMPORT FOREIGN SCHEMA "CitiesInf"  FROM SERVER fgdbtest  INTO fgdbq;


-- will import all tables --
DROP SCHEMA IF EXISTS fgdbq CASCADE;
CREATE SCHEMA fgdbq;


IMPORT FOREIGN SCHEMA ogr_all  
    FROM SERVER fgdbtest  INTO fgdbq;







```

**Comment by robe2** [comment]
okay I screwed up again and put the wrong number on my pull request. Should have been linked to this ticket - https://github.com/pramsey/pgsql-ogr-fdw/pull/47

**Comment by robe2** [comment]
Okay the EXCEPT / LIMIT TO I confirmed from Tom Lane is handled by the post processor, and it does filter out CREATE FOREIGN TABLE statements that are in violation of the LIMIT TO / EXCEPT.

So that explains when I set  laundering_table_names 'false'  it worked as expected.  The whole driver specific thing was more to do with the funkiness of names.

I think we should still direclty handle LIMIT/EXCEPT ourselves first (like postgres_fdw does) just because depending on datasource, getting the whole table definition might be expensive.

I changed the logic I had and the readme to reflect that the LIMIT TO / EXCEPT should NOT use the layer names (unless table name laundering is off), but should use post laundered names.  That does mean users have to have an idea how the laundering works though.
